File: signed-header-requests/signedRequest.m
#include <CommonCrypto/CommonDigest.h>
#include <CommonCrypto/CommonHMAC.h>

NSDictionary *makeSignedAuthHeader(
    NSString *endpoint,
    NSDictionary *params,
    NSString *datetime,
    NSString *clientId,
    NSString *secret
) {
    NSArray *paramKeys = [params allKeys];
    NSArray *sortedParamKeys = [paramKeys sortedArrayUsingSelector:@selector(compare:)];
    NSMutableArray *kvParams = [NSMutableArray array];
    for (NSString *key in sortedParamKeys) {
        [kvParams addObject:[NSString stringWithFormat:@"%@=%@", key, params[key]]];
    }
    NSString *paramString = [kvParams componentsJoinedByString:@"\n"];
    NSString *stringToSign = [NSString stringWithFormat:@"%@\n%@\n%@\n", endpoint, datetime, paramString];
    const char *cKey = [secret cStringUsingEncoding:NSASCIIStringEncoding];
    const char *cStringToSign = [stringToSign cStringUsingEncoding:NSASCIIStringEncoding];
    unsigned char cHMAC[CC_SHA1_DIGEST_LENGTH];
    CCHmac(kCCHmacAlgSHA1, cKey, strlen(cKey), cStringToSign, strlen(cStringToSign), cHMAC);
    NSData *HMAC = [[NSData alloc] initWithBytes:cHMAC length:sizeof(cHMAC)];
    NSString *signature = [HMAC base64EncodedStringWithOptions:NSDataBase64Encoding64CharacterLineLength];
    NSArray *authHeaderPieces = [NSArray arrayWithObjects:clientId,signature, nil];
    NSDictionary *authHeader = @{
        @"Authorization" : [@"Signature " stringByAppendingString:[authHeaderPieces componentsJoinedByString:@":"]]
    };
    return authHeader;
}


NSDictionary *params = @{
    @"entity_type" : @"user",
    @"filter" : @"lastUpdated >= '2016-01-01'"
};
NSDictionary *authHeader = makeSignedAuthHeader(
    @"/entity.find",
    params,
    @"2016-02-26 19:08:44",
    @"apkrahlfumwse2e9nvrrotv6vchuptzw",
    @"rylicq8ydkz0vmki3gqaoxbk4gyrr05t"
);

NSLog(@"%@", authHeader);

File: signed-header-requests/SignedRequest.java
import java.security.InvalidKeyException;
import java.security.Key;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

public class SignedRequest {
  private static final String HMAC_SHA1_ALGORITHM = "HmacSHA1";

  public static Map<String, String> makeSignedAuthHeader(
    String endpoint,
    Map<String, String> params,
    String datetime,
    String clientId,
    String secret)
  throws NoSuchAlgorithmException, InvalidKeyException {
    Set<String> paramKeys = new TreeSet<String>(params.keySet());
    List<String> kvParams = new ArrayList<String>();
    for (String key : paramKeys) {
      kvParams.add(String.format("%s=%s", key, params.get(key)));
    }
    String paramString = String.join("\n", kvParams);
    String stringToSign = String.format("%s\n%s\n%s\n", endpoint, datetime, paramString);
    Key signingKey = new SecretKeySpec(secret.getBytes(), HMAC_SHA1_ALGORITHM);
    Mac hmacInstance = Mac.getInstance(HMAC_SHA1_ALGORITHM);
    hmacInstance.init(signingKey);
    byte[] rawHmac = hmacInstance.doFinal(stringToSign.getBytes());
    String signature = new String(Base64.getEncoder().encode(rawHmac));
    Map<String, String> result = new HashMap<String,String>();
    result.put("Authorization", String.format("Signature %s:%s", clientId, signature));
    return result;
  }

  public static void main(String[] args)
  throws NoSuchAlgorithmException, InvalidKeyException {
    Map<String, String> params = new HashMap<String, String>();
    params.put("entity_type", "user");
    params.put("filter", "lastUpdated >= '2016-01-01'");

    System.out.println(makeSignedAuthHeader(
      "/entity.find",
      params,
      "2016-02-26 19:08:44",
      "apkrahlfumwse2e9nvrrotv6vchuptzw",
      "rylicq8ydkz0vmki3gqaoxbk4gyrr05t"
    ));
  }
}
